core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1377,17 +1377,27 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     def sendResponseCallback(memberState: Array[Byte], error: Errors) {
       sendResponseMaybeThrottle(request, requestThrottleMs =>
-        new SyncGroupResponse(requestThrottleMs, error, ByteBuffer.wrap(memberState)))
+        new SyncGroupResponse(
+          new SyncGroupResponseData()
+            .setErrorCode(error.code())
+            .setAssignment(memberState)
+            .setThrottleTimeMs(requestThrottleMs)
+        ))
     }
 
-    if (!authorize(request.session, Read, Resource(Group, syncGroupRequest.groupId(), LITERAL))) {
+    if (!authorize(request.session, Read, Resource(Group, syncGroupRequest.data.groupId, LITERAL))) {
       sendResponseCallback(Array[Byte](), Errors.GROUP_AUTHORIZATION_FAILED)
     } else {
+      val assignmentMap = immutable.Map.newBuilder[String, Array[Byte]]
+      syncGroupRequest.data.assignments().asScala.foreach { assignment =>
```

Comment:
nit: unneeded parenthesis `assignments()`. A few more below too.

Probably not a big deal here, but the protocol generation allows us to specify a "map key." This can be used to automatically generate a type which can be queried like a map. The magic is to add this to the schema definition:
```json
       { "name": "MemberId", "type": "string", "versions": "0+",
         "about": "The ID of the member to assign.", "mapKey": true },
```
With this, we can save some of these copies. The only problem is that it doesn't give you an actual map and the raw type is more cumbersome. We could probably improve this by offering an immutable Map view, but we don't have to do it here. I will discuss with @cmccabe whether this is a good idea or not.

Reply:
Thank you for the context! Yep, does seem like a new feature we want to address on protocol level first.